src/common/mfu_io.c

```diff
@@ -449,10 +449,50 @@ int mfu_lstat64(const char* path, struct stat64* buf)
     return rc;
 }
 
-/* use a noop, daos does not have a mknod function */
+/* emulates mknod with dfs_open, dfs_release.
+ * dfs_open supports regular files, symlinks, and directories */
 int daos_mknod(const char* path, mode_t mode, dev_t dev, mfu_file_t* mfu_file)
 {
-    return 0;
+#ifdef DAOS_SUPPORT
+    char* name     = NULL;
+    char* dir_name = NULL;
+    parse_filename(path, &name, &dir_name);
+    assert(dir_name);
+
+    dfs_obj_t* parent = NULL;
+    int rc = dfs_lookup(mfu_file->dfs, dir_name, O_RDWR, &parent, NULL, NULL);
+    if (parent != NULL) {
+        /* create regular file */
+        rc = dfs_open(mfu_file->dfs, parent, name,
+                      S_IFREG | mode, O_CREAT,
```

Comment:
I know @dsikich mentioned adding ``S_IFREG`` was needed here but now I forget why.  Can you also add a comment about that here so I don't forget again?

Reply:
@adammoody if you don't specify S_IFREG then dfs_open will return with an error saying "invalid inputs." It expects you to tell it if it is a regular file.

Reply:
Thanks @dsikich 
I'll check on this for my own information and leave an appropriate comment.

Reply:
@daltonbohning yeah if you go tohttps://github.com/daos-stack/daos/blob/master/src/client/dfs/dfs.c in dfs_open it has a switch statement that checks the file type. If it doesn't match then it returns EINVAL, maybe there is a way around passing it in explicitly, not %100 sure on that.

Reply:
(Please let me know if any of this doesn't make sense)

So the call to `mfu_file_mknod` within `mfu_flist_copy` actually passes in `S_IFREG` directly:
https://github.com/daltonbohning/mpifileutils/blob/cb2d35c20662569dc1385f375fe17e267b898b90/src/common/mfu_flist_copy.c#L1130

But the definition for `mfu_file_open` does not have a `mode` parameter directly. Instead, `mode` can be passed to `mfu_file_open` through "variable arguments", and not all calls to `mfu_file_open` pass a `mode`, and hence do not pass `S_IFREG`, which is why `daos_open` uses `S_IFREG | mode`, in case it was not passed in.
https://github.com/daltonbohning/mpifileutils/blob/cb2d35c20662569dc1385f375fe17e267b898b90/src/common/mfu_io.c#L706

`mfu_open` checks for variable arguments, and passes `mode` to `open` only if `O_CREAT` is specified, as is required by `open`:
https://man7.org/linux/man-pages/man2/open.2.html
If `mode` is not passed to `open`, and the file already exists, it doesn't matter to POSIX, because it will open the file of whatever type it is.

But in DAOS, `dfs_open` DOES require `mode` with the filetype, even if the file already exists.

POSIX `mknod`, however, always requires `mode` to be passed in to specify the file type.

But I wonder if doing `S_IFREG | mode` might be a problem if `mode = S_IFDIR` is passed into `daos_open`?
Since `dfs_open` only supports `S_IFREG`, `S_IFDIR`, `S_IFLNK`, maybe `daos_mknod` and/or `daos_open` should more appropriately have something like this (but in a better-coded way):
```
switch (mode & S_IFMT) {
	case S_IFREG:
	        // good
	case S_IFDIR:
	        // good
	case S_IFLNK:
	        // good
	default:
                set mode = S_IFREG
```
This would allow any of `S_IFREG`, `S_IFDIR`, `S_IFLNK` to be be passed to `daos_open` or `daos_mknod`, and would only default `mode` if neither are passed in.

Reply:
Update: `dfs_open` requires `value` to be set if the type is `S_IFLINK`, so maybe we would only allow `S_IFREG`, `S_IFDIR` for now.

Reply:
@daltonbohning not sure why we would ever call dfs_open on a directory? we should be using daos_mkdir, daos_opendir for those operations I would think.

Reply:
In that case, I think the `S_IFREG` would then be the only one allowed, as it is now.
And, as Adam suggested previously, I think I'll add a check to make sure that's the only type passed to `mknod`, and leave `daos_open` with the `S_IFREG | mode`, as you originally put in.
(But I'll leave some comments so we don't need this discussion again)

Reply:
@daltonbohning yeah that sounds good